src/ZPublisher/xmlrpc.py

```diff
@@ -10,15 +10,209 @@
+
+    def __delattr__(self, name):
+        return delattr(self._real, name)
+
+    def setBody(self, body, title='', is_error=0, bogus_str_search=None):
+        if isinstance(body, xmlrpclib.Fault):
+            # Convert Fault object to XML-RPC response.
+            body = xmlrpclib.dumps(body, methodresponse=1, allow_none=True)
+        else:
+            # Marshall our body as an XML-RPC response. Strings will be sent
+            # strings, integers as integers, etc. We do *not* convert
+            # everything to a string first.
+            # Previously this had special handling if the response
+            # was a Python None. This is now patched in xmlrpclib to
+            # allow Nones nested inside data structures too.
+            try:
+                body = xmlrpclib.dumps(
+                    (body,), methodresponse=1, allow_none=True)
+            except ConflictError:
+                raise
+            except:
+                self.exception()
+                return
+        # Set our body to the XML-RPC message, and fix our MIME type.
+        self._real.setBody(body)
+        self._real.setHeader('content-type', 'text/xml')
+        return self
+
+    def exception(self, fatal=0, info=None,
+                  absuri_match=None, tag_search=None):
+        # Fetch our exception info. t is type, v is value and tb is the
+        # traceback object.
+        if isinstance(info, tuple) and len(info) == 3:
+            t, v, tb = info
+        else:
+            t, v, tb = sys.exc_info()
+
+        # Don't mask 404 respnses, as some XML-RPC libraries rely on the HTTP
+        # mechanisms for detecting when authentication is required. Fixes Zope
+        # Collector issue 525.
+        if issubclass(t, Unauthorized):
+            return self._real.exception(fatal=fatal, info=info)
+
+        # Create an appropriate Fault object. Containing error information
+        Fault = xmlrpclib.Fault
+        f = None
+        try:
+            # Strip HTML tags from the error value
+            vstr = str(v)
+            remove = [r"<[^<>]*>", r"&[A-Za-z]+;"]
+            for pat in remove:
+                vstr = re.sub(pat, " ", vstr)
+            if getConfiguration().debug_mode:
+                from traceback import format_exception
+                value = '\n' + ''.join(format_exception(t, vstr, tb))
+            else:
+                value = '%s - %s' % (t, vstr)
+            if isinstance(v, Fault):
+                f = v
+            elif isinstance(v, Exception):
+                f = Fault(-1, 'Unexpected Zope exception: %s' % value)
+            else:
+                f = Fault(-2, 'Unexpected Zope error value: %s' % value)
+        except ConflictError:
+            raise
+        except Exception:
+            f = Fault(-3, "Unknown Zope fault type")
+
+        # Do the damage.
+        self.setBody(f)
+        self._real.setStatus(200)
+
+        return tb
+
+
+response = Response
```

Comment:
Is this module a wholesale clone of `ZServer.ZPublisher.xmlrpc`?  Are there any changes?  What should we do about that module?

Reply:
The only changes are the two Python 3 compatibility fixes I mentioned above.

I was also wondering what to do about the copy in ZServer. With these changes as is, it is no longer used. Maybe replace it with a deprecated imports from `ZPublisher.xmlrpc`? But that only works if both ZServer and Zope are updated, and it's hard to enforce that since neither one can have a hard dependency (i.e. with a version spec) on the other. Maybe just replace it with a module that raises an explanation explaining what happened?